urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/c/TypeDef.java

```diff
@@ -0,0 +1,441 @@
+            }
+            options.add(meth);
+        }
+
+        // Search my parents' methods, unless this is a constructor since those aren't inherited.
+        if (options.isEmpty() && !name.equals("init"))
+            for (ParameterizedType parent : parentsWithGenerics(typeGenerics))
+                try {
+                    TypeDef parentDef = Project.singleton.resolve(parent.rawType);
+                    MethodDef meth = parentDef.getInstanceMethod(
+                            name, parent.genericArgs, methGenerics, argTypes, ctx);
+                    options.add(meth);
+                } catch (NoSuchElementException e) {}
+
+        // There should be exactly one matching method.
+        if (options.isEmpty())
+            throw new NoSuchElementException(String.format(
+                    "%s has no instance method named %s accepting argument types %s",
+                    desc, name, Arrays.toString(argTypes)));
+        if (options.size() > 1)
+            throw new RuntimeException(desc + " has multiple instance methods matching " + name);
+        return options.iterator().next();
+    }
+
+    private Set<FieldDef> allInheritedFields() {
+        Set<FieldDef> fieldDefs = new HashSet<FieldDef>();
+        fieldDefs.addAll(Arrays.asList(instanceFieldDefs));
+        for (ParameterizedType parent : parents) {
+            TypeDef parentDef = Project.singleton.resolve(parent.rawType);
+            fieldDefs.addAll(parentDef.allInheritedFields());
+        }
+        return fieldDefs;
+    }
+
+    private Set<String> allInheritedFieldNames() {
+        Set<FieldDef> fieldDefs = allInheritedFields();
+        Set<String> names = new HashSet<String>();
+        for (FieldDef fieldDef : fieldDefs) {
+            String name = fieldDef.name;
+            if (names.contains(name))
+                throw new NiftyException("%s inherits multiple fields named '%s'.", desc, name);
+            names.add(name);
+        }
+        return names;
+    }
+
+    // Fetch all instance methods which this type inherits, including overridden methods.
+    // These are used to generate keys for the virtual method descriptor table.
+    private Set<MethodDef> allInheritedInstanceMethods() {
+        Set<MethodDef> methods = new HashSet<MethodDef>();
+        for (MethodDef m : instanceMethodDefs)
+            methods.add(m);
+        for (ParameterizedType parent : parents) {
+            TypeDef parentDef = Project.singleton.resolve(parent.rawType);
+            methods.addAll(parentDef.allInheritedInstanceMethods());
+        }
+        return methods;
+    }
+
+    private boolean hasSupertype(RawType sup) {
+        try {
+            thisType().asSupertype(sup, new CodeContext(this, null));
+            return true;
+        } catch (IllegalArgumentException e) {
+            return false;
+        }
+    }
+
+    // Find my implementation of an instance method.
+    private MethodDef getImplementation(MethodDef method) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 262-268, lines 280-286, lines 288-296 into separate methods.